californium-core/src/main/java/org/eclipse/californium/core/CoapObserveRelation.java

```diff
@@ -171,7 +171,7 @@ protected void setCanceled(boolean canceled) {
 			setReregistrationHandle(null);
 		}
 
-		if (notificationListener != null) {
+		if (this.canceled && notificationListener != null) {
```

Comment:
Move up two lines inside of the existing

    if (this.canceled) {
       setRegistrationhandle(null)
    }

block?